### PR TITLE
[Feature] Support AnimateDiff, a popular text2animation method

### PR DESCRIPTION
**Pull request by ElliotQi**
## Motivation

Support AnimateDiff in MMagic.

## Modification
- [x] Finish the whole pipeline to implement AnimateDiff method.
- [x] Support xformers to save memory which enables us to generate 16 frames 512\*512 videos with ~12GB VRAM.
- [x] Improve the quality of 512\*512 videos.
- [x] Finish unit tests for modules of AnimateDiff
- [x] Create a Gradio demo to make AnimateDiff easier to use.
- [ ] Train a new Motion Module based on Stable Diffusion XL.
- [ ] Support video generation based on a picture.

## Use cases (Optional)

1. Download [ToonYou](https://civitai.com/api/download/models/78775) and MotionModule checkpoint

```bash
#!/bin/bash

mkdir models && cd models
mkdir Motion_Module && mkdir DreamBooth_LoRA
gdown 1RqkQuGPaCO5sGZ6V6KZ-jUWmsRu48Kdq -O Motion_Module/
gdown 1ql0g_Ys4UCz2RnokYlBjyOYPbttbIpbu -O models/Motion_Module/
wget https://civitai.com/api/download/models/78775 -P DreamBooth_LoRA/ --content-disposition --no-check-certificate
```

2. Modify the config file in `configs/animatediff/animatediff_ToonYou.py`

```python
    models_path = {Your Checkpoints Path}
    motion_module_cfg=dict(
        path={Your MotionModule Path}
    ),
    dream_booth_lora_cfg=dict(
        type='ToonYou',
        path={Your Dreambooth_Lora Path},
        steps=25,
        guidance_scale=7.5)
```

3. Enjoy Text2Animation world

```python
from mmengine import Config

from mmagic.registry import MODELS
from mmagic.utils import register_all_modules

import os
import torch
from pathlib import Path
import datetime
from mmagic.models.editors.animatediff import save_videos_grid



register_all_modules()

cfg = Config.fromfile('configs/animatediff/animatediff_ToonYou.py')
animatediff = MODELS.build(cfg.model).cuda()
prompts = [
    "best quality, masterpiece, 1girl, looking at viewer, blurry background, upper body, contemporary, dress",

    "masterpiece, best quality, 1girl, solo, cherry blossoms, hanami, pink flower, white flower, spring season, wisteria, petals, flower, plum blossoms, outdoors, falling petals, white hair, black eyes,",

    "best quality, masterpiece, 1boy, formal, abstract, looking at viewer, masculine, marble pattern",

    "best quality, masterpiece, 1girl, cloudy sky, dandelion, contrapposto, alternate hairstyle,"
]

negative_prompts = [
    "",
    "badhandv4,easynegative,ng_deepnegative_v1_75t,verybadimagenegative_v1.3, bad-artist, bad_prompt_version2-neg, teeth",
    "",
    "",
]

sample_idx = 0
random_seeds = cfg.randomness['seed']
random_seeds = [random_seeds] if isinstance(random_seeds, int) else list(random_seeds)
samples = []
time_str = datetime.datetime.now().strftime("%Y-%m-%dT%H-%M-%S")
savedir = f"samples/{Path(cfg.model['dream_booth_lora_cfg']['type']).stem}-{time_str}"
os.makedirs(savedir)
for prompt_idx, (prompt, n_prompt, random_seed) in enumerate(zip(prompts, negative_prompts, random_seeds)):
    output_dict = animatediff.infer(prompt,negative_prompt=n_prompt, video_length=16, height=512, width=512, seed=random_seed,num_inference_steps=cfg.model['dream_booth_lora_cfg']['steps'])
    sample = output_dict['samples']
    prompt = "-".join((prompt.replace("/", "").split(" ")[:10]))
    save_videos_grid(sample, f"{savedir}/sample/{sample_idx}-{prompt}.gif")
    print(f"save to {savedir}/sample/{prompt}.gif")
    samples.append(sample)
    sample_idx += 1

samples = torch.concat(samples)
save_videos_grid(samples, f"{savedir}/sample.gif", n_rows=4)


```

## Checklist

Submitting this pull request means that,

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmagic/blob/main/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmagic/blob/main/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by ElliotQi** [comment]
## A demo with ToonYou style

ToonYou results

**Review by LeoXing1996 — CHANGES_REQUESTED** [review]
Hey @ElliotQi , thanks you for support this COOL algorithm to MMagic.
Please install pre-commit hook via pip install pre-commit before creating commit in order to fix lint error automatically.

Since you have already create a commit and PR, you can use the following commands to roll back:
```
git reset --soft HEAD^   # reset commit and remain changes
pip install pre-commit   # install pre-commit hook package
pre-commit install  # install pre-commit hook for current repo
git commit -m "xxxxx"  # create commit
git push -f <repository> <refspec>
```

**Comment by ElliotQi** [comment]
Hi @LeoXing1996, thanks for your reply.
Apart from [first commit for animatediff](https://github.com/open-mmlab/mmagic/pull/1980/commits/15d09e4c7e8d26a8f6e49be851dfcce83b6fe117), all the commits including the latest one [delete test_animatediff.py](https://github.com/open-mmlab/mmagic/pull/1980/commits/016c86efd36f42a397f398144e9c3a4fb1323604) are built after pre-commit. 
I find that all the listings are passed in the CircleCI, but the docstring coverage rate doesn't reach the minimum rate(90%).
I'll commit after adding docstrings.

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @ElliotQi,

We'd like to express our appreciation for your valuable contributions to the mmagic. Your efforts have significantly aided in enhancing the project's quality.
It is our pleasure to invite you to join our community thorugh Discord_Special Interest Group (SIG) channel. This is a great place to share your experiences, discuss ideas, and connect with other like-minded people. To become a part of the SIG channel, send a message to the moderator, OpenMMLab, briefly introduce yourself and mention your open-source contributions in the #introductions channel. Our team will gladly facilitate your entry. We eagerly await your presence. Please follow this link to join us: ​https://discord.gg/UjgXkPWNqA.

If you're on WeChat, we'd also love for you to join our community there. Just add our assistant using the WeChat ID: openmmlabwx. When sending the friend request, remember to include the remark "mmsig + Github ID".

Thanks again for your awesome contribution, and we're excited to have you as part of our community!

File: configs/animatediff/README.md [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+from pathlib import Path
+import datetime
+from mmagic.models.editors.animatediff import save_videos_grid
+
+
+
+register_all_modules()
+
+cfg = Config.fromfile('configs/animatediff/animatediff_ToonYou.py')
+animatediff = MODELS.build(cfg.model).cuda()
```
Comment (liuwenran) [inline configs/animatediff/README.md:83]: image

when I run these codes, I get this error above.
this is because If  'runwayml/stable-diffusion-v1-5' weights downloaded from huggingface automaticly, there is no diffusion_pytorch_model.bin (only diffusion_pytorch_model.safetensors). So you should add some codes to make it work with safetensors.

File: mmagic/models/editors/animatediff/attention_3d.py [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+                timestep=None,
+                attention_mask=None,
+                video_length=None):
+        """forward with hidden states, context and timestep."""
+        # SparseCausal-Attention
+        norm_hidden_states = (
+            self.norm1(hidden_states, timestep)
+            if self.use_ada_layer_norm else self.norm1(hidden_states))
+
+        # pdb.set_trace()
```
Comment (liuwenran) [inline mmagic/models/editors/animatediff/attention_3d.py:383]: some unused comments like this should be deleted.

File: configs/animatediff/README.md [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+<!-- [IMAGE] -->
```
[the remark this answers is not in the document]
Comment (ElliotQi) [reply configs/animatediff/README.md:15]: I've added an image in the README.

File: configs/animatediff/animatediff_ToonYou.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
+# config for model
```
[the remark this answers is not in the document]
Comment (ElliotQi) [reply configs/animatediff/animatediff_ToonYou.py:1]: I've revised this in all configs.

File: configs/animatediff/README.md (hunk 1 above)
Comment (ElliotQi) [reply configs/animatediff/README.md:85]: I've added automatic weights downloading and successfully loaded safetensors.

File: configs/animatediff/README_zh-CN.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
+# AnimateDiff (2023)
```
[the remark this answers is not in the document]
Comment (ElliotQi) [reply configs/animatediff/README_zh-CN.md:1]: I've made them aligned.

File: configs/animatediff/README.md [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    - "blur, haze, deformed iris, deformed pupils, semi-realistic, cgi, 3d, render, sketch, cartoon, drawing, anime, mutated hands and fingers, deformed, distorted, disfigured, poorly drawn, bad anatomy, wrong anatomy, extra limb, missing limb, floating limbs, disconnected limbs, mutation, mutated, ugly, disgusting, amputation"
+    - "blur, haze, deformed iris, deformed pupils, semi-realistic, cgi, 3d, render, sketch, cartoon, drawing, anime, art, mutated hands and fingers, deformed, distorted, disfigured, poorly drawn, bad anatomy, wrong anatomy, extra limb, missing limb, floating limbs, disconnected limbs, mutation, mutated, ugly, disgusting, amputation"
+
+```
+
+### Using MMInferencer
+
```
[the remark this answers is not in the document]
Comment (ElliotQi) [reply configs/animatediff/README.md]: I also think this could be finished later for the first video inferencer.

File: mmagic/models/editors/animatediff/attention_3d.py (hunk 2 above)
Comment (ElliotQi) [reply mmagic/models/editors/animatediff/attention_3d.py]: I've deleted unused comments.

File: configs/animatediff/README.md
Comment (liuwenran) [inline configs/animatediff/README.md:1]: An instruction should be added to demo/readme.md. like this:
https://github.com/open-mmlab/mmagic/blob/main/demo/README.md#312-vico

File: configs/animatediff/README.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+from pathlib import Path
+import datetime
+from mmagic.models.editors.animatediff import save_videos_grid
+
+
+
+register_all_modules()
+
+cfg = Config.fromfile('configs/animatediff/animatediff_ToonYou.py')
+animatediff = MODELS.build(cfg.model).cuda()
```
Comment (liuwenran) [inline configs/animatediff/README.md:81]: There is another bug when I run this script.
image

File: mmagic/models/editors/animatediff/animatediff_utils.py [hunk 8]
```diff
@@ -0,0 +1,1106 @@
+# Copyright 2023 The HuggingFace Inc. team.
```
Comment (liuwenran) [inline mmagic/models/editors/animatediff/animatediff_utils.py:1]: These comments should be removed. Make it just like your resnet_3d.py and unet_block.py

File: configs/animatediff/README.md (hunk 7 above)
Comment (ElliotQi) [reply configs/animatediff/README.md:81]: I've updated this for better adaptation in higher-version transformers.

File: mmagic/models/editors/animatediff/animatediff_utils.py (hunk 8 above)
Comment (ElliotQi) [reply mmagic/models/editors/animatediff/animatediff_utils.py]: I've removed this.

File: configs/animatediff/README.md
Comment (ElliotQi) [reply configs/animatediff/README.md:1]: I've added that in demo/readme.md

File: mmagic/models/editors/animatediff/attention_3d.py [hunk 9]
```diff
@@ -0,0 +1,699 @@
...
+
+        self.scale = dim_head**-0.5
+
+        self.heads = heads
+        # for slice_size > 0 the attention score computation
+        # is split across the batch axis to save memory
+        # You can set slice_size with `set_attention_slice`
+        self.sliceable_head_dim = heads
+        self._slice_size = None
+        self._use_memory_efficient_attention_xformers = True
```
Comment (liuwenran) [inline mmagic/models/editors/animatediff/attention_3d.py:470]: this lead to the error below.
image

File: configs/animatediff/README.md [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+    - "semi-realistic, cgi, 3d, render, sketch, cartoon, drawing, anime, text, close up, cropped, out of frame, worst quality, low quality, jpeg artifacts, ugly, duplicate, morbid, mutilated, extra fingers, mutated hands, poorly drawn hands, poorly drawn face, mutation, deformed, blurry, dehydrated, bad anatomy, bad proportions, extra limbs, cloned face, disfigured, gross proportions, malformed limbs, missing arms, missing legs, extra arms, extra legs, fused fingers, too many fingers, long neck"
+    - "blur, haze, deformed iris, deformed pupils, semi-realistic, cgi, 3d, render, sketch, cartoon, drawing, anime, mutated hands and fingers, deformed, distorted, disfigured, poorly drawn, bad anatomy, wrong anatomy, extra limb, missing limb, floating limbs, disconnected limbs, mutation, mutated, ugly, disgusting, amputation"
+    - "blur, haze, deformed iris, deformed pupils, semi-realistic, cgi, 3d, render, sketch, cartoon, drawing, anime, art, mutated hands and fingers, deformed, distorted, disfigured, poorly drawn, bad anatomy, wrong anatomy, extra limb, missing limb, floating limbs, disconnected limbs, mutation, mutated, ugly, disgusting, amputation"
+
+```
+
```
Comment (liuwenran) [inline configs/animatediff/README.md:190]: please add your training tutorial here, and put your pretrained ckpt at https://openxlab.org.cn/models.

File: mmagic/models/editors/animatediff/attention_3d.py (hunk 9 above)
Comment (ElliotQi) [reply mmagic/models/editors/animatediff/attention_3d.py:470]: I've modified this to run inference without xformers. But it's still highly recommended to install xformers for better memory cost.

File: configs/animatediff/README.md (hunk 10 above)
Comment (ElliotQi) [reply configs/animatediff/README.md:198]: I've added pretrained ckpt here and will submit new pr to support webvid dataset and relative transform.

**Review by liuwenran — APPROVED** [review]
Thanks for your contribution!
